src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ContextAware.cs

```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using System.Runtime.Serialization.DataContracts;
+
+namespace System.Runtime.Serialization
+{
+    internal sealed class ContextAwareDataContractIndex
+    {
+        private (DataContract? strong, WeakReference<DataContract>? weak)[] _contracts;
+        private ConditionalWeakTable<Type, DataContract> _keepAlive;
+
+        public int Length => _contracts.Length;
+
+        public ContextAwareDataContractIndex(int size)
+        {
+            _contracts = new (DataContract?, WeakReference<DataContract>?)[size];
+            _keepAlive = new ConditionalWeakTable<Type, DataContract>();
+        }
+
+        public DataContract? GetItem(int index) => _contracts[index].strong ?? (_contracts[index].weak?.TryGetTarget(out DataContract? ret) == true ? ret : null);
+
+        public void SetItem(int index, DataContract dataContract)
+        {
+            // Check for unloadability to decide how to store the value
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(dataContract.UnderlyingType.Assembly);
+            if (alc == null || !alc.IsCollectible)
+            {
+                _contracts[index].strong = dataContract;
+            }
+            else
+            {
+                _contracts[index].weak = new WeakReference<DataContract>(dataContract);
+                _keepAlive.Add(dataContract.UnderlyingType, dataContract);
+            }
+        }
+
+        public void Resize(int newSize)
+        {
+            Array.Resize<(DataContract?, WeakReference<DataContract>?)>(ref _contracts, newSize);
+        }
+    }
+
+    internal sealed class ContextAwareDictionary<TKey, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] TValue>
+        where TKey : Type
+        where TValue : class?
+    {
+        private readonly ConcurrentDictionary<TKey, TValue> _fastDictionary = new();
+        private readonly ConditionalWeakTable<TKey, TValue> _collectibleTable = new();
+
+
+        internal TValue GetOrAdd(TKey t, Func<TKey, TValue> f)
+        {
+            TValue? ret;
+
+            // The fast and most common default case
+            if (_fastDictionary.TryGetValue(t, out ret))
+                return ret;
+
+            // Common case for collectible contexts
+            if (_collectibleTable.TryGetValue(t, out ret))
+                return ret;
```

Comment:
Is this safe to do outside of a lock?

Reply:
I found [the documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-7.0#thread-safety) that says it is thread safe.

Reply:
The docs say it is - [ConditionalWeakTable - Thread Safety](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-7.0#thread-safety)

I looked at the code for CWT quickly and see it's using an object hash to index into fixed size arrays of entries which appear to be a self-rolled linked-list. I didn't dig into all the accessing and writing of these linked lists.

This is the slow branch of the code. I can move this inside the lock if we feel better about that.

Reply:
So, TryGetValue may be safe outside the lock here... but the GetValue() method following here executes the `createValueCallback` outside of it's lock which presents the same race that we have with ConcurrentDictionary. I'll update this.

Reply:
FWIW, the 4.8 source for CWT also makes strong claims about thread safety. https://referencesource.microsoft.com/#mscorlib/system/runtime/compilerservices/ConditionalWeakTable.cs,41